src/github_repo_request_local.py

```diff
@@ -101,27 +220,74 @@ def list_test_files(directory, excluded_extensions):
                 capture_output=True,
             )
             logger.info(f"Successfully cloned the repo: {repo_name}")
+
+            # Fetch the last commit hash and store it in the DataFrame
+            last_commit_hash = get_last_commit_hash(clone_dir)
+            df.at[index, "last_commit_hash"] = last_commit_hash
+
+            # On successful clone, increment the processed_count
+            processed_count += 1
+
         except subprocess.CalledProcessError as e:
             logger.error(
                 f"Failed to clone the repo: {repo_name}." f" " f"Exception: {e}"
             )
             df.at[index, "testfilecountlocal"] = 0
+
+            # Even on failure, consider it processed for this batch
+            processed_count += 1
+
             continue
 
-    test_file_names = list_test_files(clone_dir, args.exclude)
-    count = len(test_file_names)
-    df.at[index, "testfilecountlocal"] = count
-    formatted_names = "\n".join(test_file_names)
-    logger.info(f"Test file names for {repo_name}:\n{formatted_names}")
+    # Always attempt to fetch the last commit hash if not already fetched
+    if pd.isna(row["last_commit_hash"]):
+        last_commit_hash = get_last_commit_hash(clone_dir)
+        df.at[index, "last_commit_hash"] = last_commit_hash
 
-    # Save after each update
+    # Count test files if not already counted
+    if row["testfilecountlocal"] == -1:
+        test_file_names = list_test_files(clone_dir, args.exclude)
+        count = len(test_file_names)
+        df.at[index, "testfilecountlocal"] = count
+        formatted_names = "\n".join(test_file_names)
+        logger.info(f"Test file names for {repo_name}:\n{formatted_names}")
```

Comment:
What does this output for projects with lots of test files? I get the impression it'll be a very large log message. If so, it might be more suitable to have an option e.g. `--verbose-results=True` that'd control whether to output the list of the test names, and these would either be written to a filename based on the repo's URL, or perhaps to STDOUT (effectively output to the terminal screen but could be redirected to a text file when calling this script). LMK if additional explanation/discussion would be helpful on this. I don't know how familiar you'll be with these concepts and practices.

Reply:
Could you please confirm if you mean the `--verbose-results=True` be a command-line argument (like the ones we have using argparse`?)

I just read about the STDOUT. I think it's  a good idea to have them written in a file instead of the python console screen.

I believe later on we will need the content of these test files to be able to do analysis on them. How about saving the r`epourl`s as well as the test file names in that url? shall I save them all into one file?

Reply:
Enhance test file logging and specify output file path.

- Redirect stdout to `repo_testfiles_output.txt` for persistent logging of discovered test files.
- Save the output file in the structured data directory to maintain organisation and ease of access.
- Updated the script's `docstring` to accurately describe these changes and their impact on the script's functionality.